Avoid redundant checks in cycle detection

fcanas opened the pull request:
Dramatically improve speed of cycle detection for large or well-connected manifest graphs.

### Motivation:

Loading a project with a large or well-connected package graph in Xcode can be very slow. 

Xcode uses `Workspace.loadPackageGraph(rootPath:...)`, which calls into
calls into [`findCycle()`](https://github.com/apple/swift-package-manager/blob/25c671ef3ef2bd7a245403f9848cb3be3321c790/Sources/PackageGraph/PackageGraph%2BLoading.swift#L866). `findCycle` has a bug leading to slower than linear cycle detection.

### Modifications:

There is a note to convert to use an algorithm tracking traversal with a stack, which would allow for linear time cycle detection. But given the output of `findCycle` is the path to first the cycle, and the cycle itself, not all the elements in a strongly connected subgraph, it's enough to memoize fully explored nodes.

Once all of a Manifests outbound edges have been traversed for cycles, it's not possible for it to participate in a cycle. So if a fully-explored Manifest is encountered again in a cycle check, exit early.

Without such an early exit, cycle detection is quadratic. This change makes it linear.

### Result:

I made a tool to generate and profile the loading of package graphs through the same path as Xcode [here](https://github.com/fcanas/manifest-cycle-bug).

These are the results of profiling package loading a small range of number of packages.

| Number of packages | apple/spm | memoized |
|--------------------|-----------|----------|
| 22                 | 17.9s     | 0.2s     |
| 23                 | 35.8s     | 0.2s     |
| 24                 | 88.1s     | 0.2s     |
| 25                 | 160.5s    | 0.2s     |
| 26                 | 323.4s    | 0.2s     |

MaxDesiatov commented on Sources/PackageGraph/PackageGraph+Loading.swift:
```diff
@@ -869,13 +869,21 @@ fileprivate func findCycle(
     // Ordered set to hold the current traversed path.
     var path = OrderedCollections.OrderedSet<Manifest>()
+	
+	var fullyVisitedManifests = Set<Manifest>()
```
formatting looks inconsistent for this line

fcanas replied:
Thanks! Should be fixed now.

MaxDesiatov commented:
@swift-ci please smoke test

neonichu commented:
@swift-ci please smoke test linux

fcanas commented:
This change can also apply cleanly to the 5.7 branch. Given the impact, I would hope this would be considered for consideration in a future release of Xcode.

fcanas commented:
@tomerd I'm happy to add tests.

I'm guessing I should add fixtures somewhere in [Fixtures/DependencyResolution](https://github.com/apple/swift-package-manager/tree/main/Fixtures/DependencyResolution). Do you have any guidance as to whether this is "External" or "Internal"? And do you have any preference on the naming?

The tests will otherwise look like what I've shared in https://github.com/fcanas/manifest-cycle-bug, except with fixed fixtures to follow the existing pattern. Unless you'd like me to share the fixture generation as well.

fcanas commented:
I've added a performance test following as many of the patterns I could see for such tests, including putting the test in a subclass of `XCTestCasePerf`, and using the fixture loading utility.

I'm not sure how this project manages baselines for performance measurements. If this test were to run on master, the time would be several minutes on almost any machine. On an M1 MacBook Air, it's just under 5 seconds.

I'm happy to reshape or move what I've added.

tomerd commented on Tests/WorkspaceTests/WorkspaceTests.swift:
```diff
@@ -12355,6 +12355,25 @@ final class WorkspaceTests: XCTestCase {
+		try fixture(name: "Miscellaneous/ManifestCycleDetection/Performance/") { fixturePath in
+			let rootPackagePath = fixturePath.appending(component: "Apex")
+			
+			let workspace = try! Workspace(forRootPackage: rootPackagePath)
```
try! ->  try

tomerd commented on Tests/WorkspaceTests/WorkspaceTests.swift:
```diff
@@ -12355,6 +12355,25 @@ final class WorkspaceTests: XCTestCase {
+			let N = 1
+			measure {
+				for _ in 1..<N {
+					try! workspace.loadPackageGraph(rootPath: rootPackagePath, observabilityScope: observability.topScope)
```
try!  ->  try

tomerd commented on Tests/WorkspaceTests/WorkspaceTests.swift:
```diff
@@ -12355,6 +12355,25 @@ final class WorkspaceTests: XCTestCase {
     }
 }
 
+final class WorkspacePerformanceTests: XCTestCasePerf {
```
perhaps we should move this to `PackageGraphPerfTests`? @neonichu opinions?

neonichu replied:
Yah, I think testing that at the package graph rather than the workspace level sounds right to me, but I am not sure we have infra there to load from fixtures.

tomerd replied:
`fixture` is in SPMTestSupport so should be fine? if not, we can leave it here

fcanas replied:
Once I looked at testing the package graph more directly instead of via the workspace, doing it with generated Manifests rather than fixtures seemed viable, and more in keeping with the other Package tests.

Updated.

abertelrud commented:
@swift-ci please smoke test

neonichu commented:
@swift-ci please smoke test

tomerd reviewed (approved):
thank you @fcanas

fcanas commented:
I've created a [pull request for the 5.7 branch](https://github.com/apple/swift-package-manager/pull/5847)

neonichu commented:
Interesting error from the self-hosted job:

```
Undefined symbols for architecture x86_64:
  "_$s12SPMBuildCore23ProductBuildDescriptionMp", referenced from:
      _symbolic ______p 12SPMBuildCore23ProductBuildDescriptionP in BuildOperationBuildSystemDelegateHandler.swift.o
      _symbolic Say______pG 12SPMBuildCore23ProductBuildDescriptionP in BuildOperationBuildSystemDelegateHandler.swift.o
  "_$s12SPMBuildCore23ProductBuildDescriptionPAAE6binary8TSCBasic12AbsolutePathVvg", referenced from:
      _$s5Build0A11DescriptionV4plan13swiftCommands0d8FrontendE0013testDiscoveryE00g10EntryPointE004copyE018pluginDescriptionsAcA0A4PlanC_SDySS15LLBuildManifest17SwiftCompilerToolVGSDySSAM0qfS0VGSDySSAM04TesthS0VGSDySSAM0tijS0VGSDySSAM04CopyS0VGSayAA06PluginB0CGtKcfc12SPMBuildCore05BuiltT7ProductVA4_0zaB0_pXEfU3_ in BuildOperationBuildSystemDelegateHandler.swift.o
```

neonichu commented:
@swift-ci please smoke test macOS

tomerd commented:
```
/Users/ec2-user/jenkins/workspace/swift-package-manager-with-xcode-self-hosted-PR-osx/branch-main/swiftpm/Tests/PackageGraphPerformanceTests/PackageGraphPerfTests.swift:97:61: error: unable to infer complex closure return type; add explicit type to disambiguate
        let packageSequence = try packageNumberSequence.map { sequenceNumber in
                                                            ^
                                                                             -> <#Result#> 
[18/66] Compiling PackageGraphPerformanceTests DependencyResolverPerfTests.swift
[19/67] Merging module PackagePluginAPITests
[20/67] Merging module PackageLoadingPerformanceTests
[2
```

@fcanas some of the CI machines use older compiler which needs a bit of help with type inference. should be easy to fix by adding the type signature

neonichu commented:
@swift-ci please smoke test
